pkg/crd/spec.go

```diff
@@ -0,0 +1,137 @@
+	ApplyToCRD(crd *apiext.CustomResourceDefinitionSpec, version string) error
+}
+
+// NeedCRDFor requests the full CRD for the given group-kind.  It requires
+// that the packages containing the Go structs for that CRD have already
+// been loaded with NeedPackage.
+func (p *Parser) NeedCRDFor(groupKind schema.GroupKind) {
+	p.init()
+
+	if _, exists := p.CustomResourceDefinitions[groupKind]; exists {
+		return
+	}
+
+	var packages []*loader.Package
+	for pkg, gv := range p.GroupVersions {
+		if gv.Group != groupKind.Group {
+			continue
+		}
+		packages = append(packages, pkg)
+	}
+
+	defaultPlural := inflect.NewDefaultRuleset().Pluralize(strings.ToLower(groupKind.Kind))
+	crd := apiext.CustomResourceDefinition{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: defaultPlural + "." + groupKind.Group,
+		},
+		Spec: apiext.CustomResourceDefinitionSpec{
+			Group: groupKind.Group,
+			Names: apiext.CustomResourceDefinitionNames{
+				Kind: groupKind.Kind,
+				Plural: defaultPlural,
+			},
+		},
+	}
+
+	for _, pkg := range packages {
+		typeIdent := TypeIdent{Package: pkg, Name: groupKind.Kind}
+		typeInfo := p.Types[typeIdent]
+		if typeInfo == nil {
+			continue
+		}
+		fullSchema := FlattenEmbedded(p.flattener.FlattenType(typeIdent))
+		ver := apiext.CustomResourceDefinitionVersion{
+			Name: p.GroupVersions[pkg].Version,
+			Served: true,
+			Schema: &apiext.CustomResourceValidation{
+				OpenAPIV3Schema: fullSchema,
+			},
+		}
+		crd.Spec.Versions = append(crd.Spec.Versions, ver)
+	}
+
+	// markers are applied *after* initial generation of objects
+	for _, pkg := range packages {
+		typeIdent := TypeIdent{Package: pkg, Name: groupKind.Kind}
+		typeInfo := p.Types[typeIdent]
+		if typeInfo == nil {
+			continue
+		}
+		ver := p.GroupVersions[pkg].Version
+
+		for _, markerVals := range typeInfo.Markers {
+			for _, val := range markerVals {
+				crdMarker, isCrdMarker := val.(CRDMarker)
+				if !isCrdMarker {
+					continue
+				}
+				if err := crdMarker.ApplyToCRD(&crd.Spec, ver); err != nil {
+					pkg.AddError(loader.ErrFromNode(err, /* an okay guess */ typeInfo.RawSpec))
+				}
+			}
+		}
+	}
+
+	// fix the name if the plural was changed (this is the form the name *has* to take, so no harm in chaning it).
+	crd.Name = crd.Spec.Names.Plural + "." + groupKind.Group
+
+	// nothing to actually write
+	if len(crd.Spec.Versions) == 0 {
+		return
+	}
+
+	// make sure we have *a* storage version
+	// (default it if we only have one, otherwise, bail)
+	if len(crd.Spec.Versions) == 1 {
+		crd.Spec.Versions[0].Storage = true
+	}
+
+	hasStorage := false
+	for _, ver := range crd.Spec.Versions {
+		if ver.Storage {
+			hasStorage = true
+			break
+		}
+	}
+	if !hasStorage {
+		// just add the error to the first relevant package for this CRD,
+		// since there's no specific error location
+		packages[0].AddError(fmt.Errorf("CRD for %s has no storage version", groupKind))
+	}
+
+	p.CustomResourceDefinitions[groupKind] = crd
+}
```

Comment:
Reviewed crd parsing. This is bulk of the code I guess, looks pretty good on a high level.